packages/flutter/lib/src/cupertino/tab.dart

```diff
@@ -0,0 +1,161 @@
+  final Map<String, WidgetBuilder> routes;
+
+  /// The route generator callback used when the tab is navigated to a named route.
+  ///
+  /// This is used if [routes] does not contain the requested route.
+  final RouteFactory onGenerateRoute;
+
+  /// Called when [onGenerateRoute] also fails to generate a route.
+  ///
+  /// This callback is typically used for error handling. For example, this
+  /// callback might always generate a "not found" page that describes the route
+  /// that wasn't found.
+  ///
+  /// The default implementation pushes a route that displays an ugly error
+  /// message.
+  final RouteFactory onUnknownRoute;
+
+  /// The list of observers for the [Navigator] created in this tab.
+  ///
+  /// This list of observers is not shared with ancestor or descendent [Navigator]s.
+  final List<NavigatorObserver> navigatorObservers;
+
+  @override
+  Widget build(BuildContext context) {
+    return new Navigator(
+      onGenerateRoute: _onGenerateRoute,
+      onUnknownRoute: _onUnknownRoute,
+      observers: navigatorObservers ?? const <NavigatorObserver>[], // Can't be null.
+    );
+  }
+
+  Route<dynamic> _onGenerateRoute(RouteSettings settings) {
+    final String name = settings.name;
+    WidgetBuilder builder;
+    if (name == Navigator.defaultRouteName && home != null)
+      builder = (BuildContext context) => home;
+    else if (routes != null)
+      builder = routes[name];
+    if (builder != null) {
+      return new CupertinoPageRoute<dynamic>(
+        builder: builder,
+        settings: settings,
+      );
+    }
+    if (onGenerateRoute != null)
+      return onGenerateRoute(settings);
+    return null;
+  }
+
+  Route<dynamic> _onUnknownRoute(RouteSettings settings) {
+    assert(() {
+      if (onUnknownRoute == null) {
+        throw new FlutterError(
+          'Could not find a generator for route $settings in the $runtimeType.\n'
+          'Generators for routes are searched for in the following order:\n'
+          ' 1. For the "/" route, the "home" property, if non-null, is used.\n'
+          ' 2. Otherwise, the "routes" table is used, if it has an entry for '
+          'the route.\n'
+          ' 3. Otherwise, onGenerateRoute is called. It should return a '
+          'non-null value for any valid route not handled by "home" and "routes".\n'
+          ' 4. Finally if all else fails onUnknownRoute is called.\n'
+          'Unfortunately, onUnknownRoute was not set.'
+        );
+      }
+      return true;
+    });
+    final Route<dynamic> result = onUnknownRoute(settings);
+    assert(() {
+      if (result == null) {
+        throw new FlutterError(
+          'The onUnknownRoute callback returned null.\n'
+          'When the $runtimeType requested the route $settings from its '
+          'onUnknownRoute callback, the callback returned null. Such callbacks '
+          'must never return null.'
+        );
+      }
+      return true;
+    });
+    return result;
+  }
```

Comment:
this code duplication is unfortunate. We'll have to clean this up when we do the routes refactoring.

Reply:
please file a bug with the `routes` label to remind me to do this

Reply:
Filed #12108 but didn't really come up with a good way of doing it. 

The key part is that I want public fields to have customizable dartdocs which was hard to do. 

Mixins just makes all the fields getters and not settable in the mixer's constructor. It could be a class by composition. Though it makes the MaterialApp WidgetsApp interface a bit awkward this shared logic is split between 2 classes.

Reply:
This whole API is going to be refactored in a few months anyway. I'll clean it up then. No worries.